subprojects/model-core/src/main/java/org/gradle/internal/reflect/TypeValidationContext.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.reflect;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+public interface TypeValidationContext {
+    enum Severity {
+        /**
+         * A validation warning, emitted as a deprecation warning during runtime.
+         */
+        WARNING("Warning"),
+
+        /**
+         * A validation warning only emitted for cacheable work items, emitted as a deprecation warning during runtime.
+         */
+        CACHEABLE_WARNING("Warning"),
```

Comment:
`CACHEABLE_WARNING` is also emitted for strict mode for non-cacheable tasks. I think this is a warning about things which only affect cacheability. Those warnings are emitted even for non-cacheable tasks, we only decide to ignore them when reporting.

Reply:
What we discussed was that we want to emit these warnings (they are about missing normalization) even for non-cacheable tasks, because they might be forced to be cached via `cacheIf()`.